### PR TITLE
Modules support servicelocator tree traversal. (#14184)

### PR DESCRIPTION
**Pull request by SamMousa**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | maybe
| Fixed issues  | #14184

### CODE REVIEW

**Comment by dynasource** [comment]
this approach could lead to unexpected behavior. Getting a parental component is not always preferable.

Why not introducing a new method (e.g. `getInTree`) so that the developer can choose this kind of behavior.

**Comment by samdark** [comment]
Any cases why this is not preferrable?

**Comment by SamMousa** [comment]
@dynasource In the original ticket where I proposed this that was already discussed.

The only changed behavior here is in very specific cases:
1. Your module has a parent.
2. Your module does not have a component named `X`.
3. Your parent does have a component named `X`.
4. You want to get an exception when you call `$module->get('X')`

Likely if you get an exception you either create the component yourself or manually check the parent..

**Comment by dynasource** [comment]
why should, in case of a non-existing service, a parental service with the same name be chosen be default. Such an approach is too magical i.m.o.

**Comment by SamMousa** [comment]
Name a scenario where you would not want that. A service maps a name to an
object. If the names are the same it is expected the provided services are
the same. (You wouldn't call your i18n component `db`). In case a module
requests a component currently it just doesn't work or requires explicit
reference to the main app.

**Comment by dynasource** [comment]
practice shows that there are always exceptions. I am trying to already anticipate on that fact.

**Comment by samdark** [comment]
@dynasource if there are no concrete objections, I prefer a solution that covers majority of cases not to complicate code because of case we can't even think of.

**Comment by dynasource** [comment]
perhabs.

Another thing. How do you want to implement this new convention in the `has()` part? It could be confusing to `get()` a component while `has()` returns false.

**Comment by klimov-paul** [comment]
I tend to agree with @dynasource : introduction of such behavior is inconsistent and risk prone.

Module should be designed in the way to be be independent entity and should rely on its own component or the componets from the global application.
Making fallback for the owner module component in case it is not found inside the current one may lead to developer mistakes. Developer should be noted that component he requires does not present in the module, so he can react properly. Finding of the component with the same name in parent module can hide the logic mistake from developer.
Besides who will check the found component has an appropriate type (class)? I may expect component `foo` in some module to be a DB connection, why in parent module it can be cache.

Modules are NOT inherited - they are nested. This is done in order to support route creation for the particular controller. Being nested particular module does not inherit anything from its owner and can not why it should.
If anywone wishes to create module-based-component inheritance - he should do it on his own.

**Comment by samdark** [comment]
Thanks for the input @klimov-paul. Makes sense.

**Comment by SamMousa** [comment]
> Module should be designed in the way to be be independent entity and should rely on its own component or the componets from the global application.

I disagree with this. This method of using the service locator is terribly bad. It means modules will hardcode their relation with global app components not allowing the developer that includes the module to configure it.
Basically you're saying it's better to use a global variable instead of injecting the service locator into the module. (The service locator in this case is the parent module).

> Modules are NOT inherited - they are nested. This is done in order to support route creation for the particular controller. Being nested particular module does not inherit anything from its owner and I can not see why it should.

If they are nested then their configuration should be as well. Any code directly doing stuf like:
````
\Yii::$app('i18n')->....
````
I cannot control. If however the module uses its own service locator to get the `i18n` component and that service locator falls back to its parent (which basically means it falls back to the global component) I can override that via configuration.

Directing 3rd party libraries to code against `Yii::$app` does not feel like good practice to me. 

Furthermore, since modules can be used recursively there is no way to provide them with different definitions for the same component. What if a module wants to use a database component named `db`, but I do not want it to use the same component as another module??

> Besides who will check the found component has an appropriate type (class)? I may expect component `foo` in some module to be a DB connection, why in parent module it can be cache.

This is not an argument against tree traversal, this is an argument against the service locator pattern. Who checks that a module named `i18n` or `db` is of the proper class? The whole point of a service locator is to translate a name to an implementation.
If you are not allowed to assume the parent module has component `X`, why then are you assuming `Yii::$app` has that component? These assumptions are exactly the same.

The benefits of this tree traversal is that it allows for sensible fallback so modules can access the applications' or their parents' `X` component without hard coding against the tree. Instead they just ask "their" module for the component, allowing the developer maximum control (I can configure the module) but not breaking if I choose not to provide module specific configuration.

**Comment by SamMousa** [comment]
> Another thing. How do you want to implement this new convention in the `has()` part? It could be confusing to `get()` a component while `has()` returns false.

@dynasource I propose doing that `has()` will return the answer to the question: "Can you locate a service by this name?", since that is what you want to know. 
So in this case it would return the parents' has function in case it doesn't have a definition of its own.
I agree that it would be confusing if implemented in another way.

**Comment by klimov-paul** [comment]
> This method of using the service locator is terribly bad. It means modules will hardcode their relation with global app components not allowing the developer that includes the module to configure it. Basically you're saying it's better to use a global variable instead of injecting the service locator into the module. 

I don't get it. This approach is widely use all over framework itself. And it is used so far without any complain or incedent. For example: `rbac\DbManager` configures its DB connection component as a name of the application component by default:
https://github.com/yiisoft/yii2/blob/master/framework/rbac/DbManager.php#L41-L47

Same thing for `i18n/DbMessageSource`:
https://github.com/yiisoft/yii2/blob/master/framework/i18n/DbMessageSource.php#L48-L56

and so on.
Why suddenly it becomes unacceptable?

Should we rework behavior at those components to check current module instead of global application?

> Furthermore, since modules can be used recursively there is no way to provide them with different definitions for the same component. What if a module wants to use a database component named db, but I do not want it to use the same component as another module?

Again classes like `rbac\DbManager` and `i18n\DbMessageSource` already contain the answer for that. You should referer particular component in your module in the same way as they do, creating public field or virtual property, which then can be configured to be either a standalone component, application component or other's module component.

For example:

```php
class Module extends \yii\base\Module
{
    public $db = 'db';

    public function init()
    {
        $this->db = Instance::ensure($this->db, Connection::className());
    }
}
```

Then you can configure `Module::$db` in whatever you want:

```db
'modules' => [
    'm1' => [
        'class' => 'my\module\Module',
        'db' => 'someDb',
    ],
    'm2' => [
        'class' => 'my\module\Module',
        'db' =>['class' => 'yii\db\Connection, ....'],
    ],
    'm2' => [
        'class' => 'my\module\Module',
        'db' =>function() {...},
    ],
],
```

You may as well create a virtual property, which should return particular component with your own logic around it for more flexibility:

```php
class Module extends \yii\base\Module
{
    public function getDd()
    {
        if ($this->has('db')) {
            return $this->get('db');
        }
        return Yii::$app->get('db');
    }
}
```

> This is not an argument against tree traversal, this is an argument against the service locator pattern. Who checks that a module named i18n or db is of the proper class? The whole point of a service locator is to translate a name to an implementation. If you are not allowed to assume the parent module has component X, why then are you assuming Yii::$app has that component? These assumptions are exactly the same.

They **are** different. Declaring a random component in the global application I actually extending its public interface, creating a virtual property named by this component. In general this means I create another application class, whithout actually declaring it - this is the main benfit of service locator.
Creating my own virtual application class, I write my program in assumption it uses this particular class.
It is even a common practice among the developers to add an actual `MyApplication` class to the project in order to support typehinting:

```php

class Yii extends yii\BaseYii {
    /*
     * @var MyApplication
     */
    public static $app;
}

/**
 * @property MyComponent $foo
 */
class MyApplication extends \yii\web\Application {}

Yii::$app->foo->some(); // allows typehinting here
```

Same happens for the module, especially while module class declaration is essintial here. In case module require some particular components to function it should either contain them by default or allow (and demand!) thier configuration like the example above.

Still it is absolutely fine for the module to rely on some global application components like `Request`, `Response`, `Session`, for example.

> The benefits of this tree traversal is that it allows for sensible fallback so modules can access the applications' or their parents' X component without hard coding against the tree. Instead they just ask "their" module for the component, allowing the developer maximum control (I can configure the module) but not breaking if I choose not to provide module specific configuration.

This is just a **one of the possible solutions**. It can be used by developer if he likes it - you are free to do so. But why it should be enforced by the core? 

You are still ignore my arguments about error reporting for the missing component:
https://github.com/yiisoft/yii2/pull/14350#issuecomment-311064250

If someone wishes such behavior he should add it consciously, accepting possible fallbacks.

**Comment by klimov-paul** [comment]
>The benefits of this tree traversal is that it allows for sensible fallback so modules can access the applications' or their parents' X component without hard coding against the tree. Instead they just ask "their" module for the component, allowing the developer maximum control (I can configure the module) but not breaking if I choose not to provide module specific configuration.

Here you are talking about creation of a module, which does NOT provide any information about the components and configuration it requires to function. Instead of explicit declaration of the used components you wish to rely on the parent module to provide it, while actual parent module is completely unknown in your example.
Behavior of such module is unpredictable and deprends on its place in modules chain.

**Comment by SamMousa** [comment]
Your reasoning is faulty. Depending on the global application service locator and then suggesting it is any more foolproof than on a parent service locater doesn't make sense.

But we both know your mind doesn't change. I'll put this in its own package.

**Comment by dynasource** [comment]
although I like the functionality, I still think this should not reside inside a general method `get`. This conflicts with the single responsibility of the method and it hacks the default behavior of ServiceLocator.

One can argue that it is handy and covers many 'known' usecases, but still it is magical and not obvious at forehand, clearly state by examples of @klimov-paul. My personal preference goes to software architecture of which the workings are predicatable and magic is simply not part of that. It just weird to ask a Module for a component which is not defined, and returns the component of the parent. That is simply too error prone.

There are alternative solutions which are more explicit and therefore have my choice. Either a method `getInTree` is introduced or a Module collection is implemented (Yii::$app->module->module->treeCollection()->findComponent('i18n')

**Comment by SamMousa** [comment]
@dynasource it doesnt violate single responsibility at all.
The purpose of a service locator is to locate a service... how it does that is an implementation detail from the consumer perspective.
Using `Yii::$app` in your code is a code smell and something to avoid if possible. In modules it is possible to avoid it, so you should avoid it.

**Comment by dynasource** [comment]
> The purpose of a service locator is to locate a service... how it does that is an implementation detail from the consumer perspective.
you are adding the responsibility to look in its parents. For me that is another goal and thus another responsibility.

> Using Yii::$app in your code is a code smell and something to avoid if possible. In modules it is possible to avoid it, so you should avoid it.
Isnt it obvious that it is to improve the understanding of the example given.

**Comment by samdark** [comment]
I want to get @cebe and @SilverFire opinions as well.

**Comment by SamMousa** [comment]
> Isnt it obvious that it is to improve the understanding of the example given.

I didnt mean your specific example, I meant in general.
Using `\Yii::$app` is the same as using `global $someVariable`. If it can be avoided, it should.

**Comment by klimov-paul** [comment]
> Using \Yii::$app is the same as using global $someVariable. If it can be avoided, it should.

Using `\Yii::$app` is actually called a 'Singleton' pattern, which is widely recognized and used.

**Comment by SamMousa** [comment]
Yes. As an anti pattern: https://en.m.wikipedia.org/wiki/Singleton_pattern

The only reason to have it as a Singleton in yii is because it serves as a service locator. If it didn't there would be no need to mark the application as a Singleton. It is not even really a Singleton; I can instantiate multiple application objects. It is just a global variable that references the last initialized application object.

**Comment by klimov-paul** [comment]
Listening to you - you should simply use some other PHP framework.
I have no idea how you actually withstand usage of Yii with all desgin decisions it is built on...

**Comment by SamMousa** [comment]
Ah, so to use Yii I must blindly follow instead of question decisions made?
Is it not possible to use and like Yii without agreeing with all design decisions?

**Comment by samdark** [comment]
Discussion is going into wrong direction :) Let's wait for more opinions instead.

**Comment by rob006** [comment]
This could be extracted to trait/behavior and interface - then developer could choose whether his module should inherit components or not.

**Comment by samdark** [comment]
@rob006 do you have any cases where it should not?

**Comment by klimov-paul** [comment]
> Ah, so to use Yii I must blindly follow instead of question decisions made?
Is it not possible to use and like Yii without agreeing with all design decisions?

You have your choice.
However, the architecture and philosophy Yii is built on seems to be suitable for the many developers. I suggest if it all be wrong - none would use it.

Still there are many other solutions, which follow difference practices and patterns. I find very odd that someone chooses technology, which core principles he dislikes so mych, while there are many alternatives.
This seems unproductive.

Yii is simply one of the alternatives - it is different, because being just another copy of Symfony or Laravel makes absolutly no sense as they are already an excelent examples of SOLID usage with DI coverage.

Still @samdark has a point, it is pointless to follow discusion at this direction.

**Comment by klimov-paul** [comment]
> This could be extracted to trait/behavior and interface - then developer could choose whether his module should inherit components or not.

**Comment by SamMousa** [comment]
@rob006 if it is implemented as a trait every module developer must include it, but the point is that as a module user I cannot enforce it.
To be able to actually configure it it must be implemented in the base module (it could be a configuration setting in base module that enables or disables it, there's no technical argument against it).

Then it would become:
````
'modules' => [
'someModule' => [
    'traverse' => true // default false 
]
````
It would still break BC in case that property is used for something else, but we could make that unlikely by choosing a name like `getTreeTraversal = true`.

**Comment by klimov-paul** [comment]
> This could be extracted to trait/behavior and interface - then developer could choose whether his module should inherit components or not.

That is exactly what I have said already.

I can agree with @dynasource's suggestion to introduce extra method for this:
https://github.com/yiisoft/yii2/pull/14350#issuecomment-311033537

In case everyone seems such functionality is crucial.

**Comment by SamMousa** [comment]
If it is a separate function then it doesn't help a module user at all. If I use a module I want to be able to configure how it gets its services. I do not want to change every module I use so it uses another function; that has 0 benefits.

**Comment by rob006** [comment]
@samdark I think that @klimov-paul gave some valid arguments here: https://github.com/yiisoft/yii2/pull/14350#issuecomment-311064250 

@SamMousa You don't need extra property, you could just attach behavior/event to module.

**Comment by klimov-paul** [comment]
> If it is a separate function then it doesn't help a module user at all. If I use a module I want to be able to configure how it gets its services. I do not want to change every module I use so it uses another function; that has 0 benefits.

The team has already agreed that we will not add extra magic to the DI layer, which is exactly what you demand.
You are talking so much about anti-patterns here, but when I told you your module should explicitly declare its public interface [here](https://github.com/yiisoft/yii2/pull/14350#issuecomment-311121151), you say it is not relevant.

**Comment by SamMousa** [comment]
@rob006 there are no events for service locators.
Also I've countered every argument from the comment... but i'll do it again.

> Module should be designed in the way to be be independent entity and should rely on its own component or the componets from the global application.

Actually a module should depend on its own components or those from its parents, since a module always has a parent and might be nested. Referencing the application directly breaks modules that are nested. For example:
````
'components' => [
    'db' => [
        'some database config'
    ]
],
'modules' => [
    'moduleA' => [
        'components' => [
            'db' => [
                'some other database config'
            ]
        ],
        'modules' => [
            'moduleB' => [
                'class' => ModuleB::class
            ]
        ]
]
````
Doing it this way allows the service locator of the parent to be used transparently.

> Making fallback for the owner module component in case it is not found inside the current one may lead to developer mistakes. Developer should be noted that component he requires does not present in the module, so he can react properly. Finding of the component with the same name in parent module can hide the logic mistake from developer.

Currently the logic is hidden as well since a module can just do `\Yii::$app->get()` there is no situation in which the global application doesn't error and my solution does. However my proposal allows you as an app developer to override which components the module uses.

> Besides who will check the found component has an appropriate type (class)? I may expect component foo in some module to be a DB connection, why in parent module it can be cache.

This is a generic argument against service locators not against tree traversal.

> Modules are NOT inherited - they are nested. This is done in order to support route creation for the particular controller. Being nested particular module does not inherit anything from its owner and I can not see why it should.

Then why do modules have a parent at all? Nesting creates a tree and it makes sense to traverse that tree instead of ignoring it and blindly using the root.

> If anywone wishes to create module-based-component inheritance - he should do it on his own.

This is not an argument for or against anything.

**Comment by klimov-paul** [comment]
DI already provide enough flexibility to configure a component, which includes `Closure` with any random code, which you like:

```php
[
    'db' => function() {
        // you may use tree traverse here if you want
    }
]
```

and still it is not enough for you.

> Then why do modules have a parent at all?

As I have already said [here](https://github.com/yiisoft/yii2/pull/14350#issuecomment-311064250) it is necessary for the routing, e.g. URL creation.

> Doing it this way allows the service locator of the parent to be used transparently

How this is transparent? It is magic. In your example `ModuleB` will have compoenent `db` in case its parent has it, but if it attached to the module, which have no `db` component - it will have none.
Using standalone `ModuleB` I have no clue, which components it actually requires: if I miss `db` component in the parent module, module `ModuleB` will crash, because it has no default fallback for it.

**Comment by samdark** [comment]
@rob006 Yes, the more cascading is involved, the harder to debug mistakes are. That's absolutely true. But other points @klimov-paul made are to be discussed.

> Module should be designed in the way to be be independent entity and should rely on its own component or the componets from the global application.

Doesn't "independent entity" mean that we should use module service locator as much as we can instead of relying on the application one all the time i.e. `$module->get('component_id')` instead of `Yii::$app->get('component_id')`? If so, most of the time a module would need to somehow reuse application components anyway while being able to override these. Since an application is a module itself and since when module is running its parent is usually an application it makes sense to implement nested component search as @SamMousa is suggesting.

I'd do implementation differently though, thinking about it now. Instead of doing it at module level I'd introduce nesting to service locators and DI containers in general.

> Besides who will check the found component has an appropriate type (class)? I may expect component foo in some module to be a DB connection, why in parent module it can be cache.

Module itself. Same as it happens now. I don't see any issues here **technically**.

**Comment by samdark** [comment]
> How this is transparent? It is magic. In your example ModuleB will have compoenent db in case its parent has it, but if it attached to the module, which have no db component - it will have none.
Using standalone ModuleB I have no clue, which components it actually requires: if I miss db component in the parent module, module ModuleB will crash, because it has no default fallback for it.

Crashing in this case is fine IMO. A module declared what's needed in order for it to work. If you're not providing it at any of the levels then it would crash.

**Comment by SamMousa** [comment]
That's fine with me as well @samdark I just did the minimum change for traversal; I'm definitely in favor of supporting it for all service locators. (But note that only the module level currently has the concept of parents).

**Comment by klimov-paul** [comment]
> Crashing in this case is fine IMO. A module declared what's needed in order for it to work. If you're not providing it at any of the levels then it would crash.

In this way `Module` produces a coupling with its parent, meaning it can fucntion only with some particular parent, but can not with any other.
If particular module requires to have a particular parent, then why there should be a random configuration?

**Comment by SamMousa** [comment]
> How this is transparent? It is magic. In your example ModuleB will have compoenent db in case its parent has it, but if it attached to the module, which have no db component - it will have none.
Using standalone ModuleB I have no clue, which components it actually requires: if I miss db component in the parent module, module ModuleB will crash, because it has no default fallback for it.

Actually, if the application misses a `db` component any approach will fail. Tree traversal will work if any module in the tree has a `db` component. Which means `ModuleB` nested inside `ModuleA` with a `db` component will work even if the application itself doesn't have a `db` component.

Also component configuration using closure is great, but it solves a different problem. If I configure a compoment at the module level currently it just won't be used no matter how I configure it; since the practice is for a module to use `Yii::$app->get` it won't get the component I configured.

**Comment by SamMousa** [comment]
> In this way Module produces a coupling with its parent, meaning it can fucntion only with some particular parent, but can not with any other.
If particular module requires to have a particular parent, then why there should be a random configuration?

No! It gets decoupled from the global singleton. I'm not sure you understand the idea at all... there is no error scenario where your current code will break because of this...
If you have your component in the global app it will still function!

**Comment by klimov-paul** [comment]
> If you have your component in the global app it will still function!

Who said I will have it there?

We are talking about random components here, not just DB connection. I can rely on component like `youTube`, `mailChimp` and so on.

**Comment by klimov-paul** [comment]
`ModuleB` may use component `MailChimp` for mass mail sending and expect it presence at a parent module, while being actually attached to the module, which have no such component.

**Comment by dynasource** [comment]
Looking at some of the examples above, I already foresee developers using bad coding practices like using 
 shortcuts to access global components like the 'db' with $module->module->get('db') for accessing Yii::$app->db.

**Comment by SamMousa** [comment]
Ah, so you have a magic application where nowhere you configure your `mailChimp` component but `Yii::$app->get('mailChimp');` still works? Lol

Anyway, suppose you have a component named `X` defined in your application.
Then for any module `$module` that is directly or indirectly defined in that application, `$module->get('X')` will work.


@dynasource They can do that currently, that's not related to any code I've added at all...

**Comment by SamMousa** [comment]
@klimov-paul do you understand tree traversal? From your comments it seems like you think a module will ONLY check its parent.

But since that parent also checks its parent it will traverse all the way up to the application singleton. As i've been trying to say, your existing applications won't break or need a configuration change.

**Comment by dynasource** [comment]
@SamMousa, my example should be seen in a > 2  echelon context

**Comment by SamMousa** [comment]
@dynasource I still don't get your point. Currently a module has a parent, and a module has `get()`. So what prevents anyone currently from using `$module->module->module->module->get('something')`. 

My implementation just makes it so that `$module->get('something')` will work regardless of where in the tree it's located allowing you more freedom of configuration without forcing you to repeat yourself.

**Comment by klimov-paul** [comment]
> Ah, so you have a magic application where nowhere you configure your mailChimp component but Yii::$app->get('mailChimp'); still works? Lol

You do not even listen.
[Here](https://github.com/yiisoft/yii2/pull/14350#issuecomment-311121151) I have already told you, that using some random component either at application config is an equal to declaration of my own Application class, thus my entire program works with **this** application class. And its internal code is not meant to be transfer to any other application.

You however talk about random module, which does NOT know anything about its parent or the application, but should get components from them.

**Comment by samdark** [comment]
> In this way Module produces a coupling with its parent, meaning it can function only with some particular parent, but can not with any other.

It's no different from what it was i.e. if you're using service locator as `Yii::$app->something`, you're coupling your module with application config explicitly. If you're doing it as `$module->get('something')` then you're not coupling to application. Application **may** provide fallbacks but is not required to do so anymore. One can specifiy components in the module config directly.

**Comment by samdark** [comment]
Ideally a reusable module should not ever know anything about the application. Thus, usage of application's service locator isn't good. Using module's own service locator is OK but, in reality, isn't convenient because you likely have to re-declare same config for components such as `db`. @SamMousa solution solves all that.

**Comment by samdark** [comment]
Well, even more ideally, a module should not use service locator at all but that's a different story.

**Comment by SamMousa** [comment]
> Well, even more ideally, a module should not use service locator at all but that's a different story.

And a really complicated one, so that's why I chose not to touch that :D

**Comment by klimov-paul** [comment]
OK, last attempt.

If you wish some module or component or any other class to be independent you should provide its necessary dependencies via its config.
For example, see `yii\rbac\DbManager::$db`:
https://github.com/yiisoft/yii2/blob/master/framework/rbac/DbManager.php#L41-L47
It clearly demands DB connection to be configured, while picking it from the Application instance by default.
Note, that we do NOT use `Yii::$app->get('db')` directly, keeping the ability for re-configuration.

You **may** skip such kind of configuration fields using `Yii::$app->get('db')` directly, but **only** in case of the code, which can NOT function outside of **your** particular program.

If you say `ModuleB` should be able to pick up component `massMail` from its parent `ModuleA` - this module can NOT function anywhere else except `ModuleA`.

Yii Applications have a special method named `coreComponents()` to declare all components, which are **always** available:
https://github.com/yiisoft/yii2/blob/master/framework/web/Application.php#L189-L201

Thus `yii\web\Application` can be used in several projects under the same iterface.
You can always rely it having following components:
- request
- response
- session
- user
- errorHandler

Those components are also explicitly marked as virtual properties at PHPDoc:
https://github.com/yiisoft/yii2/blob/master/framework/web/Application.php#L19-L24

If some module relies on particular component to function, it should declare a default config for it in the same way. Only then you can say that this module can be actually attached to a random owner.

**Comment by samdark** [comment]
So the only real issue with the change is that it may become harder to debug unexpected fallbacks. For example, I've declared my `db` connections in both application and module but in module I've made a typo and instead of erroring it silently falls back to the app config.

**Comment by samdark** [comment]
> If you say ModuleB should be able to pick up component massMail from its parent ModuleA - this module can NOT function anywhere else except ModuleA.

Yes but the difference is that this PR solution is able to pick component from its parent**s**, not from just a single parent. So in the end, if `ModuleA` is missing, it will looks for `massMail` in the application. Exactly the same as it does now.

**Comment by samdark** [comment]
I agree that it's better to declare dependencies either via properties as @klimov-paul suggested or even as an explicit interface in case of truly reusable module. But that doesn't mean @SamMousa solution doesn't make sense on its own.

**Comment by SamMousa** [comment]
Okay, so consider a scenario like this:
1. I have an application that uses a database `db`.
2. I have 100 modules, each of which should use a different database.
3. My application does not need to use those databases.

Does it make sense to declare components `db1` to `db100` in the application component array or does it make sense to declare a `db` component` in each module separately?
````
[
    'components' => [
        'db1' => [...],
        'db2' => [...],
        'db3' => [...],
    ],
    'modules' => [
        'A' => [
            'db' => 'db1'
        ],
        'B' => [
            'db' => 'db2'
        ],
        'C' => [
            'db' => 'db3'
        ]
`````
vs
````
[
    'components' => [
    ],
    'modules' => [
        'A' => [
            'components' => [
                'db' => [...]
            ]
        ],
        'B' => [
            'components' => [
                'db' => [...]
            ]
        ],
        'C' => [
            'components' => [
                'db' => [...]
            ]
        ]
`````

The 2nd is way cleaner, putting module configuration inside the modules' configuration and not just creating a flat list of all services in the global application.

Note that this example doesn't even require the traversal, but having the traversal will allow me to omit the configuration in case I want to use the global database.

**Comment by samdark** [comment]
@SamMousa that's bad example because it works now w/o this pull request.

**Comment by SamMousa** [comment]
I know it works, but many modules will silently ignore it and use the global service locator.

**Comment by samdark** [comment]
That's not connected to this pull request.

**Comment by SamMousa** [comment]
I'm merely trying to illustrate that the practice of public string properties is hardly implemented by third party module devs. This PR will allow them to trust their modules' service locator so they can use that instead.

**Comment by bscheshirwork** [comment]
Sorry, may be can help new functional of php?

```php

class someModule
{
    public function exposeFunction()
    {
        return Closure::fromCallable([$this, 'getDi']);
    }

    private function getDi($param)
    {
        ...
    }
}
```

**Comment by SamMousa** [comment]
@bscheshirwork we do not have a technical / coding problem; we have a problem of differing opinions.
Unfortunately those are not as easily solved.

**Comment by bscheshirwork** [comment]
May be invert case - DI not included to module and DI pass into callable throw main config?

**Comment by SamMousa** [comment]
I don't understand what you mean.
This is not about DI, it is about service locators. Exposing the DI container like you seem to propose just turns it into a service locator.

**Comment by bscheshirwork** [comment]
You propose shared service locator between Yii::$app, parent and any nested modules?

**Comment by SamMousa** [comment]
I propose that service locators in general support nesting, that way they can locate services from their parent(s).

**Comment by bscheshirwork** [comment]
So... May be can make this option is optional?
```php
[
    'components' => [
        'db1' => [...],
        'db2' => [...],
        'db3' => [...],
    ],
    'modules' => [
        'A' => [
            'db' => 'db1'
        ],
        'B' => [
            'internalServices' => ['db'],
            'components' => [
                'db' => [...] //don't merge config
            ]
        ],
        'C' => [
            'internalServices' => ['db' => 'db3'],//merge config db3 (i.e. part in main-local.php)
        ],

    ],
]
```

**Comment by SamMousa** [comment]
There is no need to use internal services explicitly in your first example.
Since there is no config merging by default.

Your second example of `internalServices` could work. Although it would cause unexpected behavior if it is not defined in the parent, if your configuration is partial that might be bad.

It is something that we definitely should look into deeper after we get this accepted.
Currently there is no traversal at all. After we have traversal, you have all the functionality you need.

After that we could look at config merging as you propose, it will reduce duplication in the config file.

**Comment by bscheshirwork** [comment]
Hm, ok. What about mark service to merge only manualy (3th example)? All of another don't merge (current)

```php
[
    'components' => [
        'db1' => [...],
        'db2' => [...],
        'db3' => [...],
    ],
    'modules' => [
        'C' => [
            'internalServices' => ['db' => 'db3'],//merge config db3 (i.e. part in main-local.php)
            'components' => [
                'db' => [...]
            ]
        ],
    ],
]
```

**Comment by SamMousa** [comment]
You're confusing merging with tree traversal. I'm not sure what your goal is? If you want to have manually merged configuration I'm all for it, but that's not (necessarily) directly related to tree traversal.

**Comment by dynasource** [comment]
I've rethought this concept. Despite being magic https://github.com/yiisoft/yii2/pull/14350#issuecomment-311291699, the main disadvantage is the potential debug https://github.com/yiisoft/yii2/pull/14350#issuecomment-311354613 issue mentioned by @samdark. The other disadvantage about coupling is countered by comment https://github.com/yiisoft/yii2/pull/14350#issuecomment-311351558

On the contrary, the developer benefits from its advantage like overriding global components with local module components. An interesting example is mentioned by @SamMousa https://github.com/yiisoft/yii2/issues/14184 about having custom i18n for a specific module. 

Because of this, module developers can remove all references to Yii::$app->get() and simply use their $module->getInTree($component) or $module->retrieve($component), which adds cleaner code, flexibility and functionality.

So I mainly see the advantages outweigh the risks

**Comment by SamMousa** [comment]
@dynasource I think we should keep using `->get()` though.

**Comment by dynasource** [comment]
sorry, I forgot to update that part. I indeed meant ->get()

**Review by dynasource — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

**Review by samdark — APPROVED** [review]
Looks good to me but I think it's lacking docs. Should be mentioned in the guide or in the module phpdoc at least.

**Comment by samdark** [comment]
@SamMousa any suggestions on where docs could be added?

**Comment by SamMousa** [comment]
We should mention here that there is tree traversal and provide an example of how to use it:
http://www.yiiframework.com/doc-2.0/guide-structure-modules.html#best-practices

We should mention here: http://www.yiiframework.com/doc-2.0/yii-di-servicelocator.html
Just above public properties.
That: Since ... Modules support tree traversal for the service locator with a link to the best practices.

We should add a paragraph / section here: http://www.yiiframework.com/doc-2.0/guide-concept-service-locator.html

That details the benefits of tree traversal in the service locator.

I propose creating a separate issue for that and adding these 3 tasks to it.
Some are rather simple, others are a bit more work. (Like best practices)

**Comment by samdark** [comment]
I'm against making it a separate issue. Such enhancement doesn't make sense without docs.

**Comment by SamMousa** [comment]
In that case I'll put it on my todo list, might need to wait till the weekend!

**Comment by SamMousa** [comment]
Unsure about whether to document how to merge configurations when we have no built-in support for it..

Might make more sense not to publish this as a best practice until we have a practice that is better than this :)


## Merging 
If you only need to have a slightly different configuration, like in the example above
where you might have 10 configuration settings for the database, but only 1 is different, you can use a `\Closure` to merge the configuration:

````
[
    'components' => [
        'db' => [
            'prefix' => 'main_',
            'setting1' => 'a',
            'setting2' => 'b'
            'setting3' => 'c'
            'setting4' => 'd'
            'setting5' => 'e'
        ]
    ],
    'modules' => [
        'mymodule' => [
            'components' => [
                // DI works, you can also use a direct reference `\Yii::$app`.
                'db' => function(Application $app) {
                    $config = $app->getComponents()['db'];
                    $config['prefix'] => 'module_';
                    return \Yii::createObject($config);
                }
            ]
        ]
    ]
]
````

File: framework/base/Module.php
Comment (cebe) [inline framework/base/Module.php]: instead of using try/catch here, I'd better call the parent implementation with `$throwException = false` and test for `null`.

**Review by cebe — CHANGES_REQUESTED** [review]
looks good to me otherwise.

**Comment by samdark** [comment]
@cebe Adjusted it.

**Review by cebe — APPROVED** [review]

**Comment by cebe** [comment]
CHANGELOG is missing.

**Comment by SamMousa** [comment]
I won't have time for this until next week; sorry!

**Comment by cebe** [comment]
No problem, we can add the changelog ourselfs ;) I want to train the @yii-bot to do that soon btw :)

**Comment by samdark** [comment]
Added changelog.

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**
